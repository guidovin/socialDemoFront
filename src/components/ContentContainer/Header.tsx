import React, { FunctionComponent, SyntheticEvent } from 'react';
import styled from "styled-components";
import { FaWindowClose } from "react-icons/fa";

const Container = styled.div`
  background-color: grey;
  display: flex;
  justify-content: flex-start;
  align-items:stretch;
  height: 13%;
  @media(max-width:450px){
    min-width:350px;
  }
`;

interface OnClickProps {
  onClick: (event: SyntheticEvent) => any;
}

const Logo = styled.div<OnClickProps>`
  font-size: 2em;
  text-overflow:ellipsis;
  white-space: nowrap;
  display:flex;
  align-items:center;
  margin:0px 20px 0px 20px;
  cursor: pointer;

  @media(max-width:450px){
    display:none;
  }
`;
const SearchBar = styled.input`
  border-width: 2px;
  border-radius: 35px;
  background-color: blue;
  padding: 0.1em 0.1em 0.1em calc(10px + 0.1em);
  font-size:2em;
  min-width:inherit;
  margin:0px 25px 0px 25px;

`;
const AlignVertical = styled.span`
  display:flex;
  align-items:center;
  min-width:140px;
`;
//example of styled function used to extend existing component
const WipeSearch = styled(FaWindowClose)<OnClickProps>`
  /* alternative way to verticaly align */
  margin-top:auto;
  margin-bottom:auto;
  font-size:2em;
  margin-left:-15px;
  margin-right:10px;
  cursor: pointer;
`
 
const Header: FunctionComponent<{ setFilter: Function, setSelectedUser: Function, selected: boolean }> = ({ setFilter, setSelectedUser, selected }) => {
  const onClick = (e: SyntheticEvent) => {
    e.preventDefault(); 
    setFilter(""); 
    setSelectedUser(null)
  };

  return(
    <Container>
      <Logo onClick={onClick}>MySocial</Logo>
      <AlignVertical>
        <SearchBar onChange={(e) => setFilter(e.target.value)} placeholder="Search"/>
      </AlignVertical>
      {selected && <WipeSearch onClick={onClick}/>}
    </Container>);
}


export default Header;