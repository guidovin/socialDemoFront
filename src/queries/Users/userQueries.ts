
import { gql, useQuery, ApolloClient, NormalizedCacheObject } from '@apollo/client';

//query fragment example
const USER_FRAG = gql`
  fragment userFrag on User {
    index
    picture
    age
    eyeColor
    company
    email
    phone
    greeting
  }
`
const FIND_QUERY = gql`
  ${USER_FRAG}
  query FIND_QUERY($name:String){
    list(name:$name) {
      # id and name could be inside fragment, they are left here as an ilustration of fragment use with other variables
      id
      name
      ...userFrag
      friends {
        id
        name
        ...userFrag
      }
    }
  }
`
export interface User {
  id: string
  index?: number
  picture?: string
  age?: string
  eyeColor?: string
  name?: string
  company?: string
  email?: string
  phone?: string
  friends?: [User]
  greeting?: string
}

//custom react-hook example using useQuery
function useFind({ name }: { name:string }): [User] | null {
  const { loading, error, data } = useQuery(FIND_QUERY, { variables: { name }, });
  if (loading) return null;
  if (error) //example error handling, could be swapped to something that handled and prevented an app crash
    throw new Error(`Failed to fetch users with query:  ${FIND_QUERY}, and variables: ${{ name }}`);
  return data.list;
}

//fetch example using apollo's client directly. for accessibility be used with WithApollo() HoF that inject client into props or apolloConsumer
async function find({ client, name } : { client: ApolloClient<NormalizedCacheObject> , name:string }){
  const { data, loading, error } = await client.query({query:FIND_QUERY, variables:{ name }});
  if(loading) return null;
  if(error) throw new Error("Failed to fetch users");
  return data.list;
}

export { 
  FIND_QUERY,
  useFind,
  find
}