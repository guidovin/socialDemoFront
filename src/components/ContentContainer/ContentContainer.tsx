import React from 'react';
import UserCard from "../../components/UserCard/UserCard";
import { FIND_QUERY, User } from "../../queries/Users/userQueries";
import { useLazyQuery, useQuery } from '@apollo/client';
import styled from "styled-components";
import Header from './Header';
import UserDetails from '../UserDetails/UserDetails';
import colors from "../../assets/css/colors";

interface ContainerProps {
  readonly size?: string;
};

const UserContainer = styled.div<ContainerProps>`
  padding:25px;
  min-width:270px;
  background-color:${colors.background};
  
  display:grid;
  grid-gap:2em;
  grid-template-columns: 1fr;
  @media (min-width: 660px) {
    grid-template-columns: 1fr 1fr
  };
  @media (min-width: 1000px) {
    grid-template-columns: 1fr 1fr 1fr
  };
  @media (min-width: 1320px) {
    grid-template-columns: 1fr 1fr 1fr 1fr
  };   
`;

const Container = styled.div`
  background-color:${colors.background};
  min-width: 332px;
`;

const NotFoundContainer = styled.span`
  font-size:2em;
  margin-left:50px;
`;

function ContentContainer() {
  const [ find, { loading:lazyLoading, data:lazyData, error:lazyError } ] = useLazyQuery(FIND_QUERY);
  const { loading:queryLoading, data:initialData, error:initialError } = useQuery(FIND_QUERY);
  //sets data to be displayd to initial data or new data from search originated from the find() function exposed by lazyQuery hook
  const data = (lazyData && !lazyLoading && !lazyError) ? lazyData : ((initialData && !lazyLoading) ? initialData : null );
  const [ filter, setFilter ] = React.useState("");
  const [ selectedUser, setSelectedUser ] = React.useState<User | null>(null);
  // previously data was loaded using the useEffect hook in react as a way to showcase useEffect usage. 
  // this introduced problems while testing and was replaced by the current cleaner version without "unexpected" async hook calls.
  // which seems to be for the better although doesnt showcase useEffect, so here is an example on its usage:
  /**
   * this hook would call find(name:filter) whenever filter was changed by the Header search bar 
   * 
   * useEffect(()=> {
   *  find({ variables : { data:filter }});
   * }, [filter])
   * 
   * */
  //simple example error and loading handlers
  if(queryLoading) return <h3>Loading...</h3>
  if(initialError) return <h3>Ops... Something went wrong</h3>
  return(
    <Container>
      <Header 
        setFilter={(name: string) => { find({ variables: { name }}); setFilter(name); }} 
        setSelectedUser={setSelectedUser} 
        selected={!!selectedUser}
      />
      { selectedUser && 
        <UserDetails user={selectedUser}/>
      }
      { !selectedUser ?
        <UserContainer data-testid="userContainer">
         {data && data.find && data.find.map((user: User) => <UserCard key={user.id+""+Date.now()+"user"} user={user} onClick={() => setSelectedUser(user)}/>)}
        </UserContainer>
                      :
        <UserContainer data-testid="friendsContainer">
          { selectedUser.friends && 
            selectedUser
              .friends
              .filter(user => user.name?.includes(filter))
              .map((user: User) => <UserCard key={user.id+""+Date.now()+"friend"} user={user} onClick={() => setSelectedUser(user)}/>)
          }
        </UserContainer>
      }
      {(!selectedUser && data?.find?.length === 0) &&<NotFoundContainer>No User Found With Search Term: {filter}</NotFoundContainer>}
    </Container>);
}

export default ContentContainer;