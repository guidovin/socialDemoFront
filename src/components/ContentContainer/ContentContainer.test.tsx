
import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { FIND_QUERY } from "../../queries/Users/userQueries"
import ContentContainer from "./ContentContainer"
import { render, waitForElement, fireEvent, screen, prettyDOM } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
// The component AND the query need to be exported
//This is a low coverage test only to ilustrate testing and mocking with the apollo client/provider
const mocks = [
  {
    request: {
      query: FIND_QUERY,

    },
    result: {
      data: {
        find: [{ 
          name: "Cecilia Phillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "123",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        },{ 
          name: "CeasdPhillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "123sdgsd",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        },{ 
          name: "Cec ps",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "124123",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        }],
      },
    },
  },
  {
    request: {
      query: FIND_QUERY,
      variables: {
        name: "Cecilia Phillips"
      }
    },
    result: () => { 
    console.log("exec0")
    return ({
      data: {
        find: [{ 
          id: "123",
          name: "Cecilia Phillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [
            { 
              name: "CeasdPhillips",
              age: "30",
              email: "ceciliaphillips@daido.com",
              id: "123sdgsd",
              index: "0",
              picture: "testStr",
              eyeColor: "testStr",
              company: "testStr",
              phone: "testStr",
              friends: [],
              greeting: "testStr",
            },{ 
              name: "Cec ps",
              age: "30",
              email: "ceciliaphillips@daido.com",
              id: "124123",
              index: "0",
              picture: "testStr",
              eyeColor: "testStr",
              company: "testStr",
              phone: "testStr",
              friends: [],
              greeting: "testStr",
            }
          ],
          greeting: "testStr",
        }],
      },
    })}
  },
]; 
 

it('Tests rendering with initial data and search query', async () => {
  await act(async ()=>{
    render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <ContentContainer/>
    </MockedProvider>)}
  );
  //checks if container (that doesnt depend on the query answer) is rendered
  const getContainer = await waitForElement(() => screen.findByTestId("userContainer"))
  expect(getContainer).toBeDefined();
  
  //checks if userCard (that depends on the query answer, so its proof of the query return) is rendered
  const getUserCard = await waitForElement(() => screen.findAllByTestId("userCard"))//example using screen object
  expect(getUserCard).toBeDefined();

  //simulates search input and checks if executed
  const searchBar = await waitForElement(() => screen.findByTestId("searchBar"))//example using callback from render destructuring
  expect(searchBar).toBeDefined();
  //awaits for event execution and wraps act() around event call to signify a side-effect producing execution
  await act(async()=> { fireEvent.input(searchBar, { target: { value: "Cecilia Phillips" } }) });


  // //gets element with data-testid equal to the search value inputed. proving that the query was executed and related data rendered
  // const getUserTestId = await waitForElement(() => screen.findByTestId("Cecilia Phillips"))
  // expect(getUserTestId).toBeDefined();
  //gets element that contains the text from the result of the search value inputed. 
  //confirming that the query was executed and related data rendered. Example of /i notation.
  const getUserName = await waitForElement(() => screen.findByText(/Cecilia Phillips/i))
  expect(getUserName).toBeDefined();
});



const snapshotMocks = [
  {
    request: {
      query: FIND_QUERY,

    },
    result: {
      data: {
        find: [{ 
          name: "Cecilia Phillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "123",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        },{ 
          name: "CeasdPhillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "123sdgsd",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        },{ 
          name: "Cec ps",
          age: "30",
          email: "ceciliaphillips@daido.com",
          id: "124123",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [],
          greeting: "testStr",
        }],
      },
    },
  },
  {
    request: {
      query: FIND_QUERY,
      variables: {
        name: "Cecilia Phillips"
      }
    },
    result: () => { 
    console.log("exec 1")
    return ({
      data: {
        find: [{ 
          id: "123",
          name: "Cecilia Phillips",
          age: "30",
          email: "ceciliaphillips@daido.com",
          index: "0",
          picture: "testStr",
          eyeColor: "testStr",
          company: "testStr",
          phone: "testStr",
          friends: [
            { 
              name: "CeasdPhillips",
              age: "30",
              email: "ceciliaphillips@daido.com",
              id: "123sdgsd",
              index: "0",
              picture: "testStr",
              eyeColor: "testStr",
              company: "testStr",
              phone: "testStr",
              friends: [],
              greeting: "testStr",
            },{ 
              name: "Cec ps",
              age: "30",
              email: "ceciliaphillips@daido.com",
              id: "124123",
              index: "0",
              picture: "testStr",
              eyeColor: "testStr",
              company: "testStr",
              phone: "testStr",
              friends: [],
              greeting: "testStr",
            }
          ],
          greeting: "testStr",
        }],
      },
    })}
  },
]; 
 
it("Tests Snapshot", async () => {
  await act(async ()=> {
    render(
      <MockedProvider mocks={snapshotMocks} addTypename={false}>
        <ContentContainer/>
      </MockedProvider>)
    }
  );

  expect(
    prettyDOM()
  ).toMatchSnapshot()
});