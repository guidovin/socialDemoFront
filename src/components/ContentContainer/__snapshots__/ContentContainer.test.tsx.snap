// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Snapshot 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-fzokOt hImtkV\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-AxhUy osugS\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"sc-AxgMl jWfLhC\\"[39m
        [36m>[39m
          [36m<b[39m
            [33mclass[39m=[32m\\"sc-fzozJi hEeWyB\\"[39m
          [36m>[39m
            [0mMySocial[0m
          [36m</b>[39m
        [36m</div>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"sc-Axmtr bIXLFR\\"[39m
        [36m>[39m
          [36m<input[39m
            [33mclass[39m=[32m\\"sc-AxheI couGmT\\"[39m
            [33mdata-testid[39m=[32m\\"searchBar\\"[39m
            [33mplaceholder[39m=[32m\\"Search\\"[39m
          [36m/>[39m
        [36m</span>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-fznZeY fJzJg\\"[39m
        [33mdata-testid[39m=[32m\\"userContainer\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"sc-AxjAm gSZlCL\\"[39m
          [33mdata-testid[39m=[32m\\"userCard\\"[39m
        [36m>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxiKw dqRfqT\\"[39m
            [33mid[39m=[32m\\"imgContainer\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"avatar loading...\\"[39m
            [36m/>[39m
          [36m</span>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxhCb jRGgxR\\"[39m
          [36m>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
              [33mdata-testid[39m=[32m\\"Cecilia Phillips\\"[39m
            [36m>[39m
              [0mName: Cecilia Phillips[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mAge: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEye color: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mCompany: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEmail: undefined[0m
            [36m</span>[39m
          [36m</span>[39m
        [36m</div>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"sc-AxjAm gSZlCL\\"[39m
          [33mdata-testid[39m=[32m\\"userCard\\"[39m
        [36m>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxiKw dqRfqT\\"[39m
            [33mid[39m=[32m\\"imgContainer\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"avatar loading...\\"[39m
            [36m/>[39m
          [36m</span>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxhCb jRGgxR\\"[39m
          [36m>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
              [33mdata-testid[39m=[32m\\"CeasdPhillips\\"[39m
            [36m>[39m
              [0mName: CeasdPhillips[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mAge: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEye color: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mCompany: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEmail: undefined[0m
            [36m</span>[39m
          [36m</span>[39m
        [36m</div>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"sc-AxjAm gSZlCL\\"[39m
          [33mdata-testid[39m=[32m\\"userCard\\"[39m
        [36m>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxiKw dqRfqT\\"[39m
            [33mid[39m=[32m\\"imgContainer\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"avatar loading...\\"[39m
            [36m/>[39m
          [36m</span>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"sc-AxhCb jRGgxR\\"[39m
          [36m>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
              [33mdata-testid[39m=[32m\\"Cec ps\\"[39m
            [36m>[39m
              [0mName: Cec ps[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mAge: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEye color: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mCompany: undefined[0m
            [36m</span>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"sc-AxirZ KGJfC\\"[39m
            [36m>[39m
              [0mEmail: undefined[0m
            [36m</span>[39m
          [36m</span>[39m
        [36m</div>[39m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
